### PR TITLE
Added VLAN fields

### PR DESCRIPTION
**Pull request by dainperkins**
added base VLAN fields

### CODE REVIEW

**Comment by dcode** [comment]
@dainperkins how do you think we should capture [Q-in-Q tagging](https://en.wikipedia.org/wiki/IEEE_802.1ad)? 

Zeek explicitly marks them as `vlan` and `inner_vlan`. 

Suricata puts two vlans in the `vlan` field, which contains a list of up to two VLANs, i.e. `[1, 10]`, where `10` would be the equivalent of `inner_vlan`. I thought maybe `network.vlan` could be a list of vlan objects, where the `mode` of the inner vlan could be `inner` or something.

I also note that you mention to expect vlan to be associated with `source` or `destination`, but from an IDS perspective, it's part of the particular flow/packet. I think in this case `network` would be a more appropriate location. Though I could see where it could fit under what you mention as well.

**Comment by dainperkins** [comment]
So I figure we're looking at a couple of scenarios:
1) metrics / config info (more details for VLAN) - e.g. SNMP from router/switch
2) just vlan number, and maybe locally significant name - most firewalls, basic netflow
3) atomic state of 802.1q tags in packets being analyzed (bro/suricata/packetbeat/wireshark) need to cover q-in-q for these.

Thinking the easiest (after chewing on this for a bit) is to add the q-in-q fields directly to network, and leave vlan.* as reusable under source/destination with specific sugggestion to only fill out the fields sent by the observer.... 

thinking being q-in-q is likely to only be reported by a packet analyzer, probably on a span port so would fall under the auspices of network.* as being information about the state of network (ethernet) tags.

more to the point - something routing/passing q-in-q is likely to handle only 1 transformation at a time (e.g. mpls router receives q-in-q  packet, reports outer vlan as source, and inner vlan as destination)

network.vlan.inner.id
network.vlan.outer.id
network.vlan.related.ids[array of q's between inner and outer if multiple are required]

I'm thinking q-in-q is unlikely to require names a single device is unlikely to be reporting on both q-in-q state as well as source & destination vlan/routing information

what do you think @dcode?

what do you think @dcode?

**Comment by neu5ron** [comment]
all, we had massive discussion on this previously :)
https://github.com/elastic/ecs/issues/318

essentially
- a field when VLAN direction (src/dst) AND or stacking/inner/outer is unknown..
- field when VLAN direction is known

**Comment by dcode** [comment]
@dainperkins I think your proposal meets my concerns and the discussion in #318 (thanks for the xref @neu5ron). Maybe just change the PR to allow the proposed structure under `network` and the proposed `inner`/`outer` specifiers, keeping the existing `source`/`destination` specifications. I think that allows for 3 scenarios you specified above.

**Comment by dainperkins** [comment]
So - @dcode & @neu5ron  

Should I add the 802.1q fields into the VLAN group as "things that can be nested under network/src/dst"

So we'd have:
**VLAN Description & Transit Information** (typical src/dst/client/server/host as well as interface)
- vlan.id
- vlan.name

VLAN Description (about the configured VLAN)
- vlan.said (10000 + vlan id)
- vlan.description
- vlan.state (active/suspended)
- vlan.mtu
- vlan.translational_bridge [0-1005, 0-1005] . (e.g. QinQ egress/ingress translations)
- vlan.mode (uni-eni, private, rspan)

network VLAN state information (802.1q-in-q specific properties of src/dst/network)
- vlan.inner.id              first .1q encapsulation
- vlan.inner.name       
- vlan.outer.id             final .1q encapsulation
- vlan.outer.name
- vlan.related.id[]        additional .1q encapsulations
- vlan.related.name[]

With notes included in descriptions about expected reuse & nesting for state of packets / hosts vs. configuration information?

**Comment by neu5ron** [comment]
yes I would note that VLAN can be nested under src/dst if applicable - no reason for it not to be nestable if completely indicative like from a router/firewall

**Comment by MikePaquette** [comment]
@dainperkins back in https://github.com/elastic/ecs/issues/318#issuecomment-564076835 you mentioned possibly allowing nesting under client/server/host as well.  I would be in favor of adding allowance for client and server nesting, but not for host.

**Comment by dainperkins** [comment]
@MikePaquette 

Originally I was thinking in e.g. a vswitch, trunked interface, etc it might be relevant at the host level (tho I suppose it might be more applicable to nest vlan under interface) 

More I think about it. the more I think its unlikely a host vlan will matter much (let alone typically be reported by anything observing traffic) but curious about your reasoning...

**Comment by dainperkins** [comment]
Another pr (https://github.com/elastic/ecs/pull/740) just brought up the consideration that perhaps ingress / egress vlan/zone/interface should be a specific set of fields under observer... as opposed to e.g. nesting e.g. vlan.id fields under network (and not adding unnecessary complexity to vlan fields - e.g. q-in-q which is more of a state of a connection than a vlan config)

**Comment by neu5ron** [comment]
I think keeping VLANs under observer only adds an even higher level of complexity for implementation. For example, use cases of logs being forwarded or outside of beats modules users of ECS will have to do multiple if/else statements to say what is the true observer.
Additionally keeping VLANs in the network schema with an option to add to observer makes a more holistic approach for true passive log sources to match data with log sources that have intimate/actual knowledge (observers).

**Comment by dainperkins** [comment]
@neu5ron should we move the VLAN Transit Information fields directly to NETWORK?  they don't seem to fit in with the rest of the VLAN configuration / state fields here (and I can't see any need to nest the q-in-q fields anywhere else, but I might be missing something)

**Comment by neu5ron** [comment]
Yeah I would put VLAN id/number, inner/outer, etc.. under network.
Then things like vlan name and configuration information under observer or other schema that makes sense. 
Is that what your asking? Sorry, I think I got confused of transit vs configuration verbiage

**Comment by dainok** [comment]
Just my 2 cents:
* if we're getting VLANs or network from a switch, it's ok to define a specific schema for VLAN 
* if we're observing packet flows from a specific passive observer (i.e. packet broker, mirror/span port), it's also ok
* if we're following a network flow (i.e. TCP flow) through a L2-3 device, we need to track ingress and egress VLAN (i.e. a router on a stick, a firewall, a load balancer....)
In the latter case I suggest to use vlan.ingress and vlan.egress on observer schema.
See also: https://github.com/elastic/ecs/pull/740

In general, working with ECS, I realize that objects can be nested (i.e. users can be within source or destination), VLANs within observer, network within source/destination, organization (AS name) within source/destination... and so on.

**Comment by dainperkins** [comment]
I think we're all on the same page.  I'll update the original interface and vlan specs to pull any ingress/egress int/zone info into observer -  I'll add a monitor one as well (span port / 1 armed sensors), put the q-in-q analysis fields under network and I'll also add the rfc info e.g.  c-tag to the description of inner and outer vlans.  

The other netflow stuff needs a little looking into (I'd like to get as much in as possible including e.g. syn, fin, rxmt, etc), and I think the calculated fields will probably need more discussion as we have yet to include any similar time based calculations...

Sorry for mixing discussion on multiple PRs - i need it to be friday

File: schemas/vlan.yml [hunk 1]
```diff
@@ -0,0 +1,137 @@
+---
+- name: vlan
+  title: VLAN
+  group: 2
+  short: Fields to describe VLAN information and configuration.
+  description: >
+    The VLAN fields describe information about the configuration of VLANs in an enterprise network, as well as 
+    the relevant 802.1q tagging information observed in network traffic.    
+
+    VLAN fields contain configuration information about virtiual LANs, with base identification
```
[the remark this answers is not in the document]
Comment (dainperkins) [reply schemas/vlan.yml]: - vlan.... would be for meta information for single encapsulated traffic
- vlan.inner/outer... would be for 2 encapsulations
- vlan.related... would be for additional encapsulations (but not the inner and outer, which would typically be the most relevant at the point of observation

Those will live under network, with the rest of the inventory items here nuder VLAN (q-in-q is not a quality of an individual vlan, but how traffic has been encapsulated at the point of observation)

Ingress, egress, & potentially monitor fields (zone / interface / vlan) would end up under observer (we could skip monitor and use only ingress for a 1-arm config for simplicity...)

I'll update for observer and network as discussed, and then we can catch up on the inventory/state fields, will try to have updated PRs in before Wednesday.

[the remark this answers is not in the document]
Comment (dainok) [reply schemas/vlan.yml]: Agree. I think vlan can be used as a compendium to describe a static object like an endpoint, an interface and so on.
VLANs of flowing packets should be within observer.

In general: is there any way to define that an object (i.e. vlan) can be nested within another one (i.e. observer) without redefining all properties?

[the remark this answers is not in the document]
Comment (webmat) [reply schemas/vlan.yml]: @dainok Yes, to give you an example from this PR, this is what the `reusable` section does [in the schema YAML](https://github.com/elastic/ecs/pull/688/files?file-filters%5B%5D=.yml#diff-77083d84955af444186cc34582a40e59R19-R24)

This leads to `vlan.*` being nested under `source` and `destination`. Most straightforward place to see this is [the csv](https://github.com/elastic/ecs/pull/688/files?file-filters%5B%5D=.csv&file-filters%5B%5D=.yml#diff-197505f4a8b25d1d9ed82f0f032faaf7).

**Comment by webmat** [comment]
I'm worried about the current approach, from the querying perspective. If there's various levels of VLAN encapsulation going on, it will be difficult to know which field to query.

Let's say I'm looking for all traffic destined for VLAN "foo1", but I don't know what the nesting situation is, I'd have to look for:

`destination.vlan.name:foo1 OR destination.vlan.inner.name:foo1 OR destination.vlan.outer.name:foo1 OR destination.vlan.related.name:foo1`

I don't have a perfect answer, but I think one simplification we can make is at the very least rename/repurpose `.related.` to `.all.`. The `vlan.all.name` fields would be meant to capture the full array of VLAN IDs and tags. The query then becomes:

`destination.vlan.all.name:foo1`

Given that, I'm wonder if we still need a "main" vlan field, and inner/outer fields... If we do, we can now hoist them out of the array to the appropriate scalar field. I suppose it's interesting to specifically query for `vlan.inner.name:...`. So I'm open to keeping the scalar fields, as we can't query for array positions.

All of the above applies to `.id` as well. I think the simplest would be to have them both be equal-length arrays:

* `*.vlan.all.id= [ "11", "12", "13" ]`
* `*.vlan.all.name= [ "foo1", "foo2", "foo3" ]`

**Comment by dainok** [comment]
> I'm worried about the current approach, from the querying perspective. If there's various levels of VLAN encapsulation going on, it will be difficult to know which field to query.

What if you allow nested VLANs? Defining vlan within vlan.
Parent VLAN is the outer, child VLAN (if available) is the inner one.

**Comment by webmat** [comment]
@dainperkins do you think it would make sense to remove the canonical fields `vlan.id` and `vlan.name`? If there's only one at play, perhaps we could recommend users put it in `vlan.outer.id|name`?

This would leave us with:

* `vlan.all.id|name`: 2 arrays of all vlan details
* `vlan.outer.id|name`: copy the appropriate outer vlan there. If only one vlan, copy it to "outer" as well.
* `vlan.inner.id|name`: when 2 or more vlans, copy the innermost one there.

**Comment by neu5ron** [comment]
so, I think going back to the original discussion may help - the reason for `vlan.inner.id`, `vlan.outer.id`, regular `vlan.id`, and `related.vlan.id` OR `vlan.related.id`
were as follows
- `vlan.inner.id`, used in stacking/q-in-q when the log source specifies this is the case
- `vlan.outer.id`, used in stacking/q-in-q when the log source specifies this is the case
- `vlan.id`, used when the log source is not aware of q-in-q (inner/outer) or source/destination of the VLAN. ie: it just logs the vlan and thats about it.
- `related.vlan.id` OR `vlan.related.id`, used to keep inline with the verbiage and documentation of the related field. Allows/solves finding a VLAN in any field.. Additionally, @webmat users could use the query `*vlan.id` OR autocomplete in Kibana helps them find VLAN fields..

hope this helps.. thanks folks

File: schemas/vlan.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (dainperkins) [reply schemas/vlan.yml]: is there a way to allow only certain fields to be nested?

[the remark this answers is not in the document]
Comment (webmat) [reply schemas/vlan.yml]: No there isn't.

[the remark this answers is not in the document]
Comment (webmat) [reply schemas/vlan.yml]: But I think removing the "inventory-type" attributes removes the need for selectively reusing, right?

If we only keep the fields needed for enrichment on network events, I think they all make sense in all the places?

**Comment by webmat** [comment]
@dainperkins with your other VLAN PR merged, we can close this, right?

**Comment by dainperkins** [comment]
@webmat - I think so

**Pull request closed without merging**